File: scala/allergies/src/main/scala/Allergies.scala
object Allergies {

  import Allergen._

  // needed for the tests to compile and run
  def apply() = this

  type Score = Int

  def isAllergicTo(allergen: Allergen.Allergen, score: Score) =
    (allergen.value & score) == allergen.value

  def allergies(score: Score): Allergens =
    AllAllergens filter (isAllergicTo(_, score))
}

object Allergen {
  type Value = Int
  type Allergens = List[Allergen]

  sealed trait Allergen { val value: Value }
  object Eggs extends Allergen { override val value = 1 }
  object Peanuts extends Allergen { override val value = 2 }
  object Shellfish extends Allergen { override val value = 4 }
  object Strawberries extends Allergen { override val value = 8 }
  object Tomatoes extends Allergen { override val value = 16 }
  object Chocolate extends Allergen { override val value = 32 }
  object Pollen extends Allergen { override val value = 64 }
  object Cats extends Allergen { override val value = 128 }

  val AllAllergens: Allergens =
    List(Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats)
}

File: haskell/allergies/Allergies.hs
module Allergies (Allergen(..), isAllergicTo, allergies) where

data Allergen =
  Eggs | Peanuts | Shellfish | Strawberries |
  Tomatoes | Chocolate | Pollen | Cats
    deriving (Enum, Eq, Show)

type Score = Int
type Allergens = [Allergen]

allergens :: Allergens
allergens = [Eggs .. Cats]

allergies :: Score -> Allergens
allergies score =
  map toAllergen allergensAndBinOne
  where
    binScore = reverse $ decToBin score
    allergensAndBinOne = filter onlyBinOne allergensAndBin
    allergensAndBin = zip allergens binScore
    toAllergen (a,_) = a

    onlyBinOne (_,1) = True
    onlyBinOne _ = False

decToBin :: Int -> [Int]
decToBin x = reverse $ decToBin' x
  where
    decToBin' 0 = []
    decToBin' y = let (a,b) = quotRem y 2 in b:decToBin' a

isAllergicTo :: Allergen -> Score -> Bool
isAllergicTo allergen =
  elem allergen . allergies
